### PR TITLE
Remove VB specialization.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
@dotnet-bot retest windows_vsi_p2_open_prtest please
// Previous failure: http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/windows_vsi_p2_open_prtest/105/
// Retest reason:

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide This PR introduces a new approach to completion item rules that removes the need for specialized baked in knowledge of VB/C# later on.  Specifically, completion providers can now explicitly state if they want their preselected completion item to be 'soft selected' or 'hard selected'  This allows certain providers (like ObjectCreation) to state that they want hard selection, but other providers (like True/False keywords) can stay soft selected.  

This makes this decision part of the data contract for completion, and removes the need for specialized language logic here later on down the pipeline.

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
retest  windows_vsi_p2_open_prtest please

**Comment by CyrusNajmabadi** [comment]
test vsi please

File: src/Features/Core/Portable/Completion/CompletionItemRules.cs [hunk 1]
```diff
@@ -47,18 +70,48 @@ public sealed class CompletionItemRules
...
+        /// <param name="enterKeyRule">Rule about whether the enter key is passed through to the editor after the selected item has been committed.</param>
+        /// <param name="formatOnCommit">True if the modified text should be formatted automatically.</param>
+        /// <param name="matchPriority">True if the related completion item should be initially selected.</param>
+        /// <returns></returns>
+        public static CompletionItemRules Create(
+            ImmutableArray<CharacterSetModificationRule> filterCharacterRules,
+            ImmutableArray<CharacterSetModificationRule> commitCharacterRules,
+            EnterKeyRule enterKeyRule,
+            bool formatOnCommit,
+            int? matchPriority)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: This is the existing factory constructor, now without optional parameters, which now forwards to the new factory constructor.

File: src/Features/Core/Portable/Completion/CompletionItemRules.cs [hunk 2]
```diff
@@ -69,25 +122,30 @@ public sealed class CompletionItemRules
...
+        /// <param name="selectionBehavior">How this item should be selected if no text has been typed after the completion list is brought up.</param>
         /// <returns></returns>
         public static CompletionItemRules Create(
             ImmutableArray<CharacterSetModificationRule> filterCharacterRules = default(ImmutableArray<CharacterSetModificationRule>),
             ImmutableArray<CharacterSetModificationRule> commitCharacterRules = default(ImmutableArray<CharacterSetModificationRule>),
             EnterKeyRule enterKeyRule = EnterKeyRule.Default,
             bool formatOnCommit = false,
-            int? matchPriority = null)
+            int? matchPriority = null,
+            CompletionItemSelectionBehavior selectionBehavior = CompletionItemSelectionBehavior.Default)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: This is the new factory constructor with all optional values.  The old factory method is there for binary compat.  But could that is recompiled will call into this constructor instead.

File: src/Features/Core/Portable/Completion/CompletionItemRules.cs [hunk 3]
```diff
@@ -6,6 +6,28 @@
...
+        Default,
+
+        /// <summary>
+        /// If no text has been typed, the item should be soft selected. This is appropriate for 
+        /// completion providers that want to provide suggestions that shouldn't interfere with 
+        /// typing.  For example a provider that comes up on space might offer items that are soft
+        /// selected so that an additional space (or other puntuation character) will not then 
+        /// commit that item.
+        /// </summary>
+        SoftSelection,
```
Comment (Pilchie) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: I'm a little surprised I didn't see something for "new" preselection in C# here.  I would expect to be able to type: `Foo f = new (` and get `Foo f = new Foo(`.

**Comment by Pilchie** [comment]
👍 except the comment about new preselection.

File: src/Features/Core/Portable/Completion/CompletionItemRules.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: Do you mean: you expected to see me talk about that case in the comment?

**Comment by CyrusNajmabadi** [comment]
I'll update these comments.

**Pull request merged**